### DISCUSSION TITLE
How can I interact with the image picker and file picker?

### DISCUSSION THREAD

**Discussion by akagupta9** [leancodepl/patrol#918] [Q&A]
Scenario :

I have one scenario of updating profile picture. To automate that I have 2 requirements -
1. Assigning camera Permission
2. Uploading profile picture from drive or gallery,, or mocking file pickers.

For requirement 1, I can see in document and we have example test also. can you please let me know for requirement 2, what all steps I have to follow, or code snippet for refrence.

**Comment by bartekpacia** [comment]
Hi @akagupta9!

I suggest you that when the file picker or gallery app is opened, you should [dump the native view hierarchy](https://patrol.leancode.co/tips-and-tricks#inspecting-native-view-hierarchy) and inspect the `window_dump.xml` file which contains the hierarchy. Once you know how the view hierarchy looks like, you can use `$.native.tap()` to navigate your way through the file pickers.

It'd be helpful if you attached some videos and screenshots of how your app and file picker/gallery look like :)

**Comment by pawlowskim** [comment]
Hello,

I was playing around with patrol for couple last days. I was not able to successful use tap method to close permissions popups (Location always allow, notifications and motion permissions). The server crashed in my case. Our main app view is map, so it found plenty of map related objects (like POI and stuff) and probably just couldn't handle it. 

So to workaround it I just added more commands to the patrol server and dart part of the lib, to find explicitly buttons by text. It worked on iOS, didn't have time to test on Android side but should work the same way. Unfortunately communication was re-written to grpc so I won't made any PR soon.

It would be lovely if someone could write more generic method to tap on alerts. Currently, it's mapping between command and text on the alert. Small change to the code that can get any text along with the command would resolve the problem and allow tap on any alerts, without need to add explicitly per type command (of course, in case, where tap method is not working).

Something like this (+ dart part,  grpc and android implementation) should work
```
func handlePermission(code: String, text: String) {
    runAction("handling permission with code \(code)") {
      let systemAlerts = self.springboard.alerts
        switch code {
        case "WHILE_USING":
            let okButton = systemAlerts.buttons["OK"]
            if okButton.exists {
                okButton.tap()
            }
            
            let allowWhileUsingButton = systemAlerts.buttons["Allow While Using App"]
            if allowWhileUsingButton.exists {
                allowWhileUsingButton.tap()
            }
        case "ONLY_THIS_TIME":
            let allowOnceButton = systemAlerts.buttons["Allow Once"]
            if allowOnceButton.exists {
                allowOnceButton.tap()
            }
        case "DENIED":
            let denyButton = systemAlerts.buttons["Don't Allow"]
            if denyButton.exists {
                denyButton.tap()
            }
        case "OK":
            let okButton = systemAlerts.buttons["OK"]
            if okButton.exists {
                okButton.tap()
            }
        case "ALWAYS_ALLOW":
            let alwaysAllowButton = systemAlerts.buttons["Change to Always Allow"]
            if alwaysAllowButton.exists {
                alwaysAllowButton.tap()
            }
        case "ALLOW":
            let allowButton = systemAlerts.buttons["Allow"]
            if allowButton.exists {
                allowButton.tap()
            }
        case "CUSTOM":
            let customButton = systemAlerts.buttons[text]
            if customButton.exists {
                customButton.tap()
            }
      default:
        Logger.shared.e("Invalid code: \(code)")
      }
    }
  }
```

**Comment by bartekpacia** [comment]
Hi @pawlowskim, thanks a lot for the feedback! We appreciate it a lot 💙

>  I was not able to successful use tap method to close permissions popups (Location always allow, notifications and motion permissions)

That's because of how the `PatrolAutomation.tap(on:inAppId)` method is implemented on iOS:

```swift
func tap(on text: String, inApp appId: String) {
    runAction("tapping on \(text)") {
      let app = XCUIApplication(bundleIdentifier: appId)
      app.descendants(matching: .any)[text].firstMatch.tap()
    }
  }
```

I guess that you were using it like this on the Dart side:

```dart
await $.native.tap(Selector(text: 'Allow'));
```

which was implicitly using the `bundleIdentifier` of your app under test (defined in `PatrolTestConfig`). The problem is that permission request dialogs don't belong to your app. They belong to the launcher, i.e `com.apple.springboard`. I'm not sure if it would work, but you could try tapping on permission request dialogs like this:

```dart
await $.native.tap(
  Selector(text: 'Allow'),
  appId: 'com.apple.springboard',
);
```

Anyway, you shouldn't have to work around the permission handling feature not working.  Thanks for your code snippet, and sorry that you encountered the bug. I'll try to fix this today :)

**Comment by pawlowskim** [comment]
Thanks for the feedback, I will try passing the launcher id for tap event. Hope it works 🤞

**Comment by bartekpacia** [comment]
@pawlowskim How'd it go? :)

**Comment by pawlowskim** [comment]
@bartekpacia AFAIR it didn't worked. But meantime, communication was rewritten to grpc and more native buttons texts were added to existing click events, and that worked.

**Comment by GentiShtjefni** [comment]
@bartekpacia I am also trying to do the same thing with trying to automate uploading an image in android. The problem I have is that I don't see any text so I can use $.native.tap, there is only some icons, firstly a camera icon to take the picture and than a tick icon to approve the photo taken. This is what I am trying to automate. Any suggestions?
![image](https://github.com/leancodepl/patrol/assets/84448161/59c73275-8414-49f2-a56f-bab761c12f31)
![image](https://github.com/leancodepl/patrol/assets/84448161/126264ac-6f4b-4863-982d-9fe7d49e6de6)

**Reply by EskinaKseniya** [reply]
Hi @GentiShtjefni! Could you please tell me if you have found the answer to your question?
I also encountered this problem and can't find a solution..

**Reply by GentiShtjefni** [reply]
Hi @EskinaKseniya! I was able to find a way to not actually do the uploading on tests. In my case I added a environment variable and some logic in the code, so when the app is in testing Im uploading a test file and dodging the whole camera app.
